source/common/router/config_impl.cc

```diff
@@ -1087,7 +1088,25 @@ RouteConstSharedPtr RouteEntryImplBase::clusterEntry(const Http::HeaderMap& head
 RouteConstSharedPtr RouteEntryImplBase::pickWeightedCluster(const Http::HeaderMap& headers,
                                                             const uint64_t random_value,
                                                             const bool ignore_overflow) const {
-  const uint64_t selected_value = random_value % total_cluster_weight_;
+  uint64_t selected_value = random_value % total_cluster_weight_;
+  // Override the selected value with provided random value if related name is specified in the
+  // header.
+  if (!random_value_name_.empty()) {
+    const auto header_value = headers.get(Envoy::Http::LowerCaseString(random_value_name_));
```

Comment:
I would probably structure this as something like 

```
uint64_t random_value_to_use;
if (random_value_name_) {
  random_value_to_use = ...
}

const uint64_t selected_value = random_value_to_use % total_cluster_weight_;
```

I think that flows a bit nicer since we avoid overwriting selected_value?

Reply:
The overwriting logic here is kinds of necessary because we want to ensure the fallback path when retrieving the value from the header failed (as shown in the unit test).

I change the code a bit by using `optional` rather than `overwriting` to reflect this